### ISSUE TITLE
NSC to close a connection with no longer existing NSE

### ISSUE THREAD

**Issue by LionelJouin** [networkservicemesh/sdk#1434]
## Expected Behavior

I would like to remove my NSE first and then close the connection from the NSC with no error. If no NSE is found for the connection, then NSM should continue to close via the forwarder so the IPs, routes, policies routes... could be removed.  

## Current Behavior

The NSC will try to close the connection, but will receive an error since one of the component in the chain (nsmgr or forwarder, I am not sure) will try to find and continue the close chain via the NSE.

I suppose this behavior is not incorrect because NSM might consider the NSE will re-appear (and re-establish the connection), so the NSC will then be able to close, and then the NSE can release resources (e.g. IPs from IPAM). But what happens if the NSE never come back?

## Failure Information (for bugs)
/

## Steps to Reproduce

1. Deploy NSC + NSE
2. Scale the NSE to 0 / Remove the NSE deployment
3. Try to close the connection from the NSC

## Context
/

## Failure Logs
/

**Comment by LionelJouin** [comment]
I tried it with only NSM, it seems this problem appears only when the HostPID is not set to true in the vpp-forwarder. @denis-tingaikin @glazychev-art Why does the vpp-forwarder need the HostPID?

**Comment by glazychev-art** [comment]
@LionelJouin 
forwarder-vpp is able to use an external vpp-instance that can be located on the node. We use it for example for Calico-vpp. Calico creates vpp-instance on the node and NSM just reuses it. We need HostPID for that.

I'm not sure that the problem is due to HostPID. Because it seems I reproduced the issue with this setting.
There is an idea how to solve this problem - https://github.com/networkservicemesh/sdk/pull/1440

**Comment by glazychev-art** [comment]
We will see errors anyway because NSE is already dead and Close will return an error.
[NSC] --- close ---> [NSMgr] ----> [Fwd] ----.... ---- x [NSE(dead)]

The main thing is that Close reaches the forwarder - the forwarder will clean it up.

**Comment by LionelJouin** [comment]
I don't see any close in the forwarder when I try to close from the NSC (When there is no more NSE running).
I found this chain element running in the vpp-forwarder.
https://github.com/networkservicemesh/sdk-vpp/blob/main/pkg/networkservice/nsmonitor/client.go
This fails to monitor without hostPID. But when it works and the NSE is removed, the monitoring is closed and it triggers the close in the vpp-forwarder that delete the interfaces, routes...

**Comment by glazychev-art** [comment]
Yes, nsmonitor is necessary for the correct Close. Looks like hostPID should be set.
Have you used this PR? https://github.com/networkservicemesh/sdk/pull/1440

**Comment by LionelJouin** [comment]
I just tried it, yes, now I can see the close going through the forwarder and the interface being removed.

**Comment by glazychev-art** [comment]
@denis-tingaikin @edwarnicke 
Any thoughts?

**Comment by LionelJouin** [comment]
About the hostPID, it seems I tried only local connection last time probably. Today I tried with a remote connection (NSC and NSE on different node), I scaled the NSEs to 0, and the interface in the NSC has not been removed, I can see the close call (due to PID monitoring) only in the forwarder-vpp where the NSE was running, is it the expected behavior?

**Comment by denis-tingaikin** [comment]
@LionelJouin Thanks for letting know. 

@glazychev-art I think we should sent the event about dying the datapath from the second forwarder? 
Could you please have a look?

**Comment by glazychev-art** [comment]
@LionelJouin 
We looked at the `hostPID` problem and Closes - all this things are related to healing.  
NSM has the ability to check the liveliness of the connection and interfaces. This feature is needed in order to optimize healing (not call Close), in case the datapath is still alive (we can send and receive data using current interfaces) but there are some other problems (for example, in the NSMgr). 

`hostPID` allows us to implement this feature - `nsmonitor` (as you correctly found out) relies on it to track the death of endpoints.

To disable liveliness checking we have to use `NSM_LIVENESSCHECKENABLED: false` env. In this case, healing will also continue to work, but `Close` will be called when any element dies (even if NSMgr restarted, which does not affect the datapath)

So we have these cases:
1. **hostPID: true**, **NSM_LIVENESSCHECKENABLED: true**:
Default behavior, the NSC instantly knows that the endpoint is dead and calls Close.

2. **hostPID: true**, **NSM_LIVENESSCHECKENABLED: false**:
Same as above, but NSC can also make unnecessary Closes (this leads to unnecessary re-creation of interfaces)

3. **hostPID: false**, **NSM_LIVENESSCHECKENABLED: true**:
We still have the feature enabled, but it doesn't work properly due to the missing hostPID flag. In this case, Closes will also be called, but by **timeout** - only after `NSM_MAX_TOKEN_LIFETIME` expires (default 10 minutes)

4. **hostPID: false**, **NSM_LIVENESSCHECKENABLED: false**:
The same as case 2.

I rechecked it on the current main branch, it's true. If your behavior is different, please attach the logs, we will try to figure it out

**Comment by LionelJouin** [comment]
Thank you for checking and for the details, I will have to check again, I don't know if I was using NSM_LIVENESSCHECKENABLED or not

**Comment by glazychev-art** [comment]
If you didn't override NSM_LIVENESSCHECKENABLED env for NSC sidecar or admissionwebhook, I guess you use this feature (it is enabled by default)
